2791 Fix handling of `onMatch` and `onMismatch` attributes in the properties configuration format

JWT007 opened the pull request:
#2791
---
+ Updated DefaultComponentBuilder `put(String, String)` to remove attributte if value is null
+ Updated DefaultComponentBuilder guarded `putAttribute(...)` implementations with object values against NPE
+ updated javadoc
+ added changelog

(edit)
+ per @ppkarwasz review added JVerify annotations to `config.builder` and `config.builder.api` packages
+ optimized builder APIs and Default***ComponentBuilder implementations
+ changed (deprecated) 'addAttribute' methods to 'setAttribute' to better match other builder APIs
+ added convencience APIs for setting general attributes on ***ComponentBuilder interfaces
+ javadoc'd both packages 100%
+ updated callers of now deprecated APIs
+ updated changelog

JWT007 committed:
Fix potential NPEs in DefaultComponentBuilder in addAttribute methods (#2791)

ppkarwasz commented on log4j-core/src/main/java/org/apache/logging/log4j/core/config/builder/impl/DefaultComponentBuilder.java:
```diff
@@ -57,43 +59,51 @@ public DefaultComponentBuilder(final CB builder, final String name, final String
     @Override
     public T addAttribute(final String key, final Enum<?> value) {
-        return put(key, value.name());
+        return put(key, Optional.ofNullable(value).map(Enum::name).orElse(null));
```
```suggestion
        return put(key, value != null ? value.name() : null);
```
We tend to use more primitive expressions in Log4j to keep it garbage-free. For a configuration it doesn't matter, but you'll find a lot of simple `null` checks and iterations on arrays in the code.

Personally I wouldn't use `Optional` for a single method call.

ppkarwasz commented on log4j-core/src/main/java/org/apache/logging/log4j/core/config/builder/impl/DefaultComponentBuilder.java:
```diff
@@ -57,43 +59,51 @@ public DefaultComponentBuilder(final CB builder, final String name, final String
         this.value = value;
     }
 
+    /** {@inheritDoc} */
```
Do we need it? Since we don't add anything, we can leave the Javadoc empty.

JWT007 replied:
oh I guess I just do that as a matter of course... I asked my AI Assistant - best practice, etc. :)

  - Including `/** {@inheritDoc} */` as a rule for all overridden methods ensures that all methods are documented in the same manner, even if no additional details are needed.
  - This can be particularly useful in libraries used by others (e.g., open-source or enterprise frameworks) where clear documentation is critical.
  -  If you're building a library or public-facing tool, adding `{@inheritDoc}` for every method provides clarity when generating documentation and ensures consistency
-  Adding `/** {@inheritDoc} */` to methods that don't add any documentation is often considered a **good practice for clarity**, especially for public APIs or widely used libraries.

ppkarwasz reviewed (changes requested):
@JWT007,

Thanks! :100:

I think we could profit from the occasion and add JSpecify annotation to all classes in this package. As far as I can tell only a few parameters and fields can be `null`, mostly those named `name` and `value`.

JWT007 committed:
Added JVerify annotations per review (#2791)

JWT007 committed:
Added DefaultComponentBuilderTest (#2791)

JWT007 committed:
Replaced DefaultComponentBuilder#put with DefaultComponentBuilder#putAttributte (#2791)

+ more appropriate method name

JWT007 committed:
Reworked the builder.api and builder.api.impl packages (#2791)

+ made all Default***Builder implementations @ProviderrType
+ added setXXXAttribute method to some component builder interfaces with default implementations
+ added JSpecify @NullMarked to packages and @Nullable where approrpiate to fields/mehods (per last PR review)
+ deprecated ComponentBuilder#addAttribute(...) methods and replaced with ComponentBuilder#setAttribute(...) methods in line with other builders (i.e. Plugins) - also since they don't just add but also remove if the value is null
+ reorganized ConfigurationBuilder to group getters, setters, builder factory methods
+ incremented package-info.java @Version annoations
+ added some null-guards in BuiltConfiguration#setup()
+ moved DeffaultConfigurationBuilder XML generation code from top to bottom of source file
+ removed "extra" attributes in Default***Builder constructors and replaced with setXXXAttribute chained calls in DefaultConfigurrationBuilder#newXXX(...)  convenience methods
+ javadoc'd both packages (fixing documentation errors)

JWT007 committed:
Update references to deprecations with new component builder APIs (#2791)

JWT007 committed:
Changed ConfigurationBuilder#setLoggerContext(...) to return builder instead of void (#2791)

+ added @BaselineIgnore due to API change (but breaks no code)

JWT007 committed:
Adapt changelog entry for additional changes (#2791)

ppkarwasz commented on log4j-core/src/main/java/org/apache/logging/log4j/core/config/builder/impl/DefaultComponentBuilder.java:
```diff
@@ -17,104 +17,217 @@
     @Override
-    public T addAttribute(final String key, final int value) {
-        return put(key, Integer.toString(value));
+    public @NonNull CB getBuilder() {
```
```suggestion
    public CB getBuilder() {
```

Isn't `@NonNull` the default?

ppkarwasz commented on log4j-core/src/main/java/org/apache/logging/log4j/core/config/builder/impl/DefaultComponentBuilder.java:
```diff
@@ -17,104 +17,217 @@
+     * </p>
+     */
+    protected void clear() {
+        synchronized (this) {
+            attributes.clear();
+            components.clear();
+        }
```
Since the class is not thread-safe anyway, do we need this?

ppkarwasz commented on log4j-core/src/main/java/org/apache/logging/log4j/core/config/builder/impl/DefaultComponentBuilder.java:
```diff
@@ -17,104 +17,217 @@
+     * Gets the list of child components.
+     * @return an <i>immutable</i> list of the child components
+     */
+    protected List<Component> getComponents() {
```
```suggestion
    protected List<? extends Component> getComponents() {
```

ppkarwasz commented on log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java:
```diff
@@ -334,7 +332,7 @@ private static <B extends ComponentBuilder<B>> ComponentBuilder<B> createCompone
         final String name = (String) properties.remove(CONFIG_NAME);
         final String type = (String) properties.remove(CONFIG_TYPE);
-        if (Strings.isEmpty(type)) {
+        if (type == null || Strings.isEmpty(type)) {
```
```suggestion
        if (Strings.isEmpty(type)) {
```

`Strings.isEmpty` is `null`-safe.

ppkarwasz commented on src/changelog/.2.x.x/2791_rework_ComponentBuilder_API.xml:
```diff
@@ -0,0 +1,12 @@
+  <issue id="2791" link="https://github.com/apache/logging-log4j2/issues/2791"/>
+  <description format="asciidoc">
+    Fix problem when null attribute values are set on DefaultComponentBuilder. GitHub issue #2791.
+    Added JVerify annotations to config.builder.* packages.
```
```suggestion
    Added JSpecify annotations to config.builder.* packages.
```

ppkarwasz reviewed (changes requested):
Thanks for the PR! :100:

Overall it looks OK, I added some notes below:

vy commented:
@ramanathan1504, is this something you might want to pick up?

ramanathan1504 commented:
@vy, Yes, I can pick this up. I will review this PR and the issue discussion and come up with updated one.

vy commented:
Superseded by #4116.
